No way to type ava's `t.context` object

## Summary: 

Ava allows users to [apply their own type](https://github.com/avajs/ava/blob/master/docs/recipes/typescript.md#typing-tcontext) to `t.context`, but there doesn't seem to be a way to thread this through `testProp` and into the `prop` function.

## Bug report:

### What I did

Created a typed `ExecutionContext` and tried to use this from within 

```typescript
import { fc, testProp } from 'ava-fast-check'
import anyTest, { TestInterface, ExecutionContext } from 'ava'

type TestContext = {
    state: string
}

const test = anyTest as TestInterface<TestContext>;

test.beforeEach(t => {
    t.context.state = 'uninitialized'
})

testProp(
    'should type `t`',
    [fc.string()],
    (t: ExecutionContext<TestContext>, s) => {
        t.pass()
    }
)
```

With the following `tsconfig.json` (this is not reproducible without `strict: true`),

```json
{
    "include": [
        "test.ts"
    ],
    "exclude": [
        "node_modules",
        "dist"
    ],
    "compilerOptions": {
        "target": "es5",
        "module": "commonjs",
        "moduleResolution": "node",
        "lib": ["es6"],
        "strict": true,
        "esModuleInterop": true,
        "outDir": "./dist"
    }
}

```

and the following package versions

```json
{
  "name": "scratch",
  "version": "1.0.0",
  "license": "ISC",
  "devDependencies": {
    "ava": "^3.13.0",
    "ava-fast-check": "^3.0.0",
    "fast-check": "^2.5.0",
    "ts-node": "^9.0.0",
    "typescript": "^4.0.3"
  },
  "ava": {
	  "file": "test.ts",
	  "extensions": ["ts"],
	  "require": ["ts-node/register"]
  }
}
```

### What I expected to happen

Test would compile and run / pass

### What actually happened

Compilation error in the type checker:

```

  Uncaught exception in test.ts

  test.ts(17,5): error TS2345: Argument of type '(t: ExecutionContext<TestContext>, s: string, n: number) => void' is not assignable to parameter of type 'Prop<[s: string, n: number]>'.
    Types of parameters 't' and 't' are incompatible.
      Type 'ExecutionContext<unknown>' is not assignable to type 'ExecutionContext<TestContext>'.
        Type 'unknown' is not assignable to type 'TestContext'.

  ✖ test.ts exited with a non-zero exit code: 1
  ─

  1 uncaught exception
```


Now that I think about it, I'm not sure that before/after hooks (where the `ExecutionContext` is normally populated) are supported at all in `ava-fast-check`, but I'm hoping they "just work". -- Update: Yes, since `test` is re-exported from `ava-fast-check`, the before/after hooks work as expected. I might document this on the readme somewhere.

P.S. this package, and fast-check, are awesome!! Thank you so much for your work on these projects

--- a/README.md
+++ b/README.md
@@ -3,10 +3,10 @@
 [![Build Status](https://travis-ci.com/dubzzz/ava-fast-check.svg?branch=master)](https://travis-ci.com/dubzzz/ava-fast-check)
 [![npm version](https://badge.fury.io/js/ava-fast-check.svg)](https://badge.fury.io/js/ava-fast-check)
 
-Bring the power of property based testing framework fast-check into ava.
-`ava-fast-check` simplifies the integration of fast-check into ava testing framework.
+Bring the power of property based testing framework fast-check into AVA.
+`ava-fast-check` simplifies the integration of fast-check into AVA testing framework.
 
-## Getting started
+## Getting Started
 
 Install `ava-fast-check` and its peer dependencies:
 
@@ -16,7 +16,7 @@ npm install --save-dev ava fast-check ava-fast-check
 
 ## Example
 
-```javascript
+```typescript
 import { testProp, fc } from 'ava-fast-check';
 
 // for all a, b, c strings
@@ -26,31 +26,88 @@ testProp('should detect the substring', [fc.string(), fc.string(), fc.string()],
 });
 ```
 
-The property is passed [`ava`'s `t` argument](https://github.com/avajs/ava/blob/master/docs/02-execution-context.md#execution-context-t-argument) for its first parameter, and the value of each arbitrary for the current test case for the rest of the parameters.
+The property is passed [AVA's `t` argument](https://github.com/avajs/ava/blob/master/docs/02-execution-context.md#execution-context-t-argument) for its first parameter, and the value of each arbitrary for the current test case for the rest of the parameters.
 
-`ava-fast-check` supports all of [`ava`'s assertions](https://github.com/avajs/ava/blob/master/docs/03-assertions.md#assertions) and like `ava`, supports synchronous and asynchronous functions, including promises, observables, and callbacks. See [`ava`'s documentation](https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#declaring-test) for more information.
+`ava-fast-check` supports all of [AVA's assertions](https://github.com/avajs/ava/blob/master/docs/03-assertions.md#assertions) and like AVA, supports synchronous and asynchronous functions, including promises, observables, and callbacks. See [AVA's documentation](https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#declaring-test) for more information.
 
 ## Advanced
 
-If you want to forward custom parameters to fast-check, `testProp` accepts an optional `fc.Parameters` ([more](https://github.com/dubzzz/fast-check/blob/master/documentation/Runners.md#runners)).
+### `fast-check` Parameters
 
-`ava-fast-check` also comes with `.only`, `.skip` and `.failing` from `ava`.
+`testProp` accepts an optional `fc.Parameters` for forwarding custom parameters to `fast-check` ([more](https://github.com/dubzzz/fast-check/blob/master/documentation/Runners.md#runners)).
 
-```javascript
+### AVA Modifiers
+
+`ava-fast-check` also comes with [`.only`], [`.serial`] [`.skip`], and [`.failing`] modifiers from AVA.
+
+```typescript
 import { testProp, fc } from 'ava-fast-check';
 
 testProp('should replay the test for the seed 4242', [fc.nat(), fc.nat()], (t, a, b) => {
   t.is(a + b, b + a);
 }, { seed: 4242 });
 
-testProp.failing('should be skipped', [fc.fullUnicodeString()], (t, text) => {
+testProp.skip('should be skipped', [fc.fullUnicodeString()], (t, text) => {
   t.is([...text].length, text.length);
 });
 ```
 
+[`.only`]: https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#running-specific-tests
+[`.serial`]: https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#running-tests-serially
+[`.skip`]: https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#skipping-tests
+[`.failing`]: https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#failing-tests
+
+### AVA `before`/`after` Hooks
+
+Import `test` from `ava-fast-check` to use `before`/`after` [hooks]:
+
+```typescript
+import { test, testProp, fc } from 'ava-fast-check';
+
+test.before(t => {
+  connectToDatabase();
+});
+
+testProp(
+  // ... omitted for brevity
+);
+
+test.after(t => {
+  closeDatabaseConnection();
+});
+```
+
+[hooks]: https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#before--after-hooks
+
+### AVA Execution Context
+
+`ava-fast-check` mirror's AVA's procedure for customizing the test [execution context]:
+
+```typescript
+import { fc, testProp as anyTestProp, PropertyTestInterface } from '../src/ava-fast-check';
+
+type TestContext = {
+  state: string
+};
+
+const testProp = anyTestProp as PropertyTestInterface<TestContext>;
+
+testProp(
+  'should reach terminal state',
+  [fc.string()],
+  (t, received) => {
+    // here t is typed as ExecutionContext<TestContext>
+    console.log(t.context.state); // logs 'uninitialized'
+    // ... omitted for brevity
+  }
+);
+```
+
+[execution context]: https://github.com/avajs/ava/blob/master/docs/02-execution-context.md
+
 ## Minimal requirements
 
-| ava-fast-check | ava                   | fast-check           |
+| ava-fast-check | AVA                   | fast-check           |
 |----------------|-----------------------|----------------------|
 | ^3.0.0         | >=3.9.0<sup>(1)</sup> | ^2.0.0<sup>(2)</sup> |
 | ^2.0.0         | >=3.9.0<sup>(1)</sup> | ^1.0.0               |

--- a/README.md
+++ b/README.md
@@ -59,12 +59,12 @@ testProp.skip('should be skipped', [fc.fullUnicodeString()], (t, text) => {
 
 ### AVA `before`/`after` Hooks
 
-Import `test` from `ava-fast-check` to use `before`/`after` [hooks]:
+`ava-fast-check` exposes AVA's `before`/`after` [hooks]:
 
 ```typescript
-import { test, testProp, fc } from 'ava-fast-check';
+import { testProp, fc } from 'ava-fast-check';
 
-test.before(t => {
+testProp.before(t => {
   connectToDatabase();
 });
 
@@ -72,7 +72,7 @@ testProp(
   // ... omitted for brevity
 );
 
-test.after(t => {
+testProp.after(t => {
   closeDatabaseConnection();
 });
 ```

--- a/package.json
+++ b/package.json
@@ -42,8 +42,8 @@
     "esm": "^3.2.25",
     "fast-check": "^2.0.0",
     "prettier": "^1.19.1",
-    "ts-node": "^8.10.2",
-    "typescript": "^3.9.6"
+    "ts-node": "^9.0.0",
+    "typescript": "^4.0.3"
   },
   "keywords": [
     "ava",

--- a/src/ava-fast-check.ts
+++ b/src/ava-fast-check.ts
@@ -1,4 +1,6 @@
 import test, {
+  AfterInterface,
+  BeforeInterface,
   ExecutionContext,
   Implementation,
   ImplementationResult,
@@ -11,9 +13,9 @@ export { fc, test };
 
 type NonEmptyArray<A> = A[] & {0: A};
 
-// Pre-requisite: https://github.com/Microsoft/TypeScript/pull/26063
-// Require TypeScript 3.1
-type ArbitraryTuple<Ts extends NonEmptyArray<any>> = { [P in keyof Ts]: fc.Arbitrary<Ts[P]> };
+type ArbitraryTuple<Ts extends NonEmptyArray<any>> = {
+  [P in keyof Ts]: fc.Arbitrary<Ts[P]>
+};
 
 type Prop<Context, Ts extends NonEmptyArray<any>> = (
   t: ExecutionContext<Context>,
@@ -33,9 +35,13 @@ type AvaModifierWhitelist =
   | 'skip'
   | 'serial'
 
-export type PropertyTestInterface<Context> = PropertyTest<Context> & {
-  [Modifier in AvaModifierWhitelist]: PropertyTest<Context>;
-}
+export type PropertyTestInterface<Context> =
+  & PropertyTest<Context>
+  & { [Modifier in AvaModifierWhitelist]: PropertyTest<Context> }
+  & {
+    before: BeforeInterface<Context>,
+    after: AfterInterface<Context>,
+  }
 
 function wrapProp<Context, Ts extends NonEmptyArray<any>>(
   arbitraries: ArbitraryTuple<Ts>,
@@ -102,6 +108,8 @@ export const testProp: PropertyTestInterface<unknown> = Object.assign(
     only: exposeModifier('only'),
     failing: exposeModifier('failing'),
     skip: exposeModifier('skip'),
-    serial: exposeModifier('serial')
+    serial: exposeModifier('serial'),
+    before: test.before,
+    after: test.after,
   }
-)
+);

--- a/src/ava-fast-check.ts
+++ b/src/ava-fast-check.ts
@@ -1,17 +1,47 @@
-import test, { ExecutionContext, Implementation, ImplementationResult, TryResult } from 'ava';
+import test, {
+  ExecutionContext,
+  Implementation,
+  ImplementationResult,
+  TestInterface,
+  TryResult
+} from 'ava';
 import * as fc from 'fast-check';
+
+export { fc, test };
+
+type NonEmptyArray<A> = A[] & {0: A};
 
 // Pre-requisite: https://github.com/Microsoft/TypeScript/pull/26063
 // Require TypeScript 3.1
-type ArbitraryTuple<Ts extends [any] | any[]> = { [P in keyof Ts]: fc.Arbitrary<Ts[P]> };
+type ArbitraryTuple<Ts extends NonEmptyArray<any>> = { [P in keyof Ts]: fc.Arbitrary<Ts[P]> };
 
-type Prop<Ts extends [any] | any[]> = (t: ExecutionContext, ...args: Ts) => ImplementationResult;
+type Prop<Context, Ts extends NonEmptyArray<any>> = (
+  t: ExecutionContext<Context>,
+  ...args: Ts
+) => ImplementationResult;
 
-function wrapProp<Ts extends [any] | any[]>(
+type PropertyTest<Context> = <Ts extends NonEmptyArray<any>>(
+  label: string,
   arbitraries: ArbitraryTuple<Ts>,
-  prop: Prop<Ts>,
+  prop: Prop<Context, Ts>,
   params?: fc.Parameters<Ts>
-): Implementation {
+) => void;
+
+type AvaModifierWhitelist =
+  | 'only'
+  | 'failing'
+  | 'skip'
+  | 'serial'
+
+export type PropertyTestInterface<Context> = PropertyTest<Context> & {
+  [Modifier in AvaModifierWhitelist]: PropertyTest<Context>;
+}
+
+function wrapProp<Context, Ts extends NonEmptyArray<any>>(
+  arbitraries: ArbitraryTuple<Ts>,
+  prop: Prop<Context, Ts>,
+  params?: fc.Parameters<Ts>
+): Implementation<Context> {
   return async t => {
     let failingTry: undefined | TryResult;
 
@@ -39,11 +69,11 @@ function wrapProp<Ts extends [any] | any[]>(
   };
 }
 
-function internalTestProp<Ts extends [any] | any[]>(
-  testFn: (label: string, exec: Implementation) => void,
+function internalTestProp<Context, Ts extends NonEmptyArray<any>>(
+  testFn: (label: string, exec: Implementation<Context>) => void,
   label: string,
   arbitraries: ArbitraryTuple<Ts>,
-  prop: Prop<Ts>,
+  prop: Prop<Context, Ts>,
   params?: fc.Parameters<Ts>
 ): void {
   const customParams: fc.Parameters<Ts> = params || {};
@@ -52,40 +82,26 @@ function internalTestProp<Ts extends [any] | any[]>(
   testFn(`${label} (with seed=${customParams.seed})`, wrapProp(arbitraries, prop, params));
 }
 
-export function testProp<Ts extends [any] | any[]>(
-  label: string,
-  arbitraries: ArbitraryTuple<Ts>,
-  prop: Prop<Ts>,
-  params?: fc.Parameters<Ts>
-): void {
-  internalTestProp(test, label, arbitraries, prop, params);
+function exposeModifier<Context, T extends Extract<keyof TestInterface, AvaModifierWhitelist>>(
+  modifier: T
+): PropertyTest<Context> {
+  return (label, arbitraries, prop, params) =>
+    internalTestProp((test as TestInterface<Context>)[modifier], label, arbitraries, prop, params);
 }
 
-export namespace testProp {
-  export const only = <Ts extends [any] | any[]>(
+export const testProp: PropertyTestInterface<unknown> = Object.assign(
+  function testProp<Context, Ts extends NonEmptyArray<any>>(
     label: string,
     arbitraries: ArbitraryTuple<Ts>,
-    prop: Prop<Ts>,
+    prop: Prop<Context, Ts>,
     params?: fc.Parameters<Ts>
-  ): void => internalTestProp(test.only, label, arbitraries, prop, params);
-  export const failing = <Ts extends [any] | any[]>(
-    label: string,
-    arbitraries: ArbitraryTuple<Ts>,
-    prop: Prop<Ts>,
-    params?: fc.Parameters<Ts>
-  ): void => internalTestProp(test.failing, label, arbitraries, prop, params);
-  export const skip = <Ts extends [any] | any[]>(
-    label: string,
-    arbitraries: ArbitraryTuple<Ts>,
-    prop: Prop<Ts>,
-    params?: fc.Parameters<Ts>
-  ): void => internalTestProp(test.skip, label, arbitraries, prop, params);
-  export const serial = <Ts extends [any] | any[]>(
-    label: string,
-    arbitraries: ArbitraryTuple<Ts>,
-    prop: Prop<Ts>,
-    params?: fc.Parameters<Ts>
-  ): void => internalTestProp(test.serial, label, arbitraries, prop, params);
-}
-
-export { test, fc };
+  ): void {
+    internalTestProp(test as TestInterface<Context>, label, arbitraries, prop, params);
+  },
+  {
+    only: exposeModifier('only'),
+    failing: exposeModifier('failing'),
+    skip: exposeModifier('skip'),
+    serial: exposeModifier('serial')
+  }
+)

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -2,12 +2,10 @@
     "compilerOptions": {
         "declaration": true,
         "sourceMap": false,
+        "strict": true,
         "alwaysStrict": true,
-        "noImplicitAny": true,
-        "noImplicitThis": true,
         "removeComments": true,
         "preserveConstEnums": true,
-        "strictNullChecks": true,
         "downlevelIteration": true,
         "module": "commonjs",
         "target": "es3",

